build_wheels.py

```diff
@@ -0,0 +1,199 @@
+# For master branch it is taken constraints file incremented by 1 in IDF minor version
+# (e.g when last release is v5.2 the constrains file for master is v5.3)
+# When no such a file exists the last release constraints file is taken for master branch
+#
+# This feature can be overwritten by setting the AUTO_VERSION_CONSTRAINS_FILE to False
+# in Github variables or manual workflow
+# then the value in MASTER_CONSTRAIN_FILE is taken
+auto_version = os.environ.get('AUTO_VERSION_CONSTRAINS_FILE', 'True')
+AUTO_VERSION_CONSTRAINS_FILE: bool = auto_version.lower() in ['true', '1']
+MASTER_CONSTRAINS_FILE: str = os.environ.get('MASTER_CONSTRAINS_FILE', 'v6.0')
+
+print(f'ENV variables: IDF v{MIN_IDF_MAJOR_VERSION}.{MIN_IDF_MINOR_VERSION}'
+      f' -- {AUTO_VERSION_CONSTRAINS_FILE} = {MASTER_CONSTRAINS_FILE}'
+      )
+
+
+def check_response(response: requests.Response, warning: str, exit_on_wrong: bool = False) -> bool:
+    """Print warning or exit the script when response code is not correct"""
+    if response.status_code == 200:
+        return True
+    if exit_on_wrong:
+        raise SystemExit(f'{warning}\n{response.text}')
+    print(warning, '\n')
+    return False
+
+
+# ESP-IDF branches list
+def fetch_idf_branches() -> List[str]:
+    """Fetch IDF branches from URL specified in global variables"""
+    res = requests.get(IDF_BRANCHES_URL, timeout=10)
+    if check_response(res, 'Failed to fetch ESP-IDF branches.', True):
+        return [branch['name'] for branch in res.json()]
+    return []
+
+
+def get_used_idf_branches(idf_repo_branches: List[str]) -> List[str]:
+    """Take only IDF master and release branches, only equal or grater version specified in Github variables"""
+    idf_branches: List[str] = []
+    for branch in idf_repo_branches:
+        idf_release = re.match(r'release/v(\d+)\.(\d+)', branch)
+
+        if not idf_release:
+            continue
+
+        idf_major, idf_minor = map(int, idf_release.groups())
+
+        if (idf_major, idf_minor) < (MIN_IDF_MAJOR_VERSION, MIN_IDF_MINOR_VERSION):
+            continue
+
+        idf_branches.append(branch)
+
+    idf_branches.append('master')
+    return idf_branches
+
+
+# Constraints files versions list
+def get_constraints_versions(idf_branches: List[str]) -> List[str]:
+    """From desired branches passed in get constraints files versions list
+    - when branch is not 'release' (without version) it is supposed to be 'master'
+    and auto version mechanism is applied if not specified in Github variables or manual workflow not to
+    """
+    idf_constraints: List[str] = []
+
+    for branch in idf_branches:
+        # Handle release branches
+        if 'release/' in branch:
+            idf_constraints.append(branch.split('release/')[1])
+            continue
+
+        # Handle master branch - NO automatic version mechanism
+        if not AUTO_VERSION_CONSTRAINS_FILE or idf_branches[0] == 'master':
```

Comment:
Hi @jakub-kocka , I'm confused by this condition. Why is the `idf_branches[0] == 'master'` part there. We know that the branches could be only `release/vX.Y` or `master`. This is assured in `get_used_idf_branches`. Also I'm not sure how `master` can be first entry in the `idf_branches` list. Wouldn't just simple
```
if not AUTO_VERSION_CONSTRAINS_FILE
```
be enough here? We know that it has to be `master` at this point right?
Thank you

Reply:
Hi, @fhrbata. This was meant to be just a failsafe when the `MIN_IDF_MAJOR_VERSION` and `MIN_IDF_MINOR_VERSION` will be set in GitHub variables to not suit any release version of IDF (grater than).

If so and the `AUTO_VERSION_CONSTRAINS_FILE` will be **left to `True`**, there will be no possible way to apply auto-mechanism followed -> uncaught Exception.

So this is a fall-back to take at least variable defined in `MASTER_CONSTRAINS_FILE` GitHub variable because `master` will be in the `idf_branches`  always **(0 position if no` release`)**.

That means `MASTER_CONSTRAINS_FILE` should have a valid constraints file version ... for now it could be 'v5.3' (`release/v5.2` is not in GitHub so the `v5.2` constraints file is taken for `master` now).

However if the `MASTER_CONSTRAINS_FILE` will not be valid, such as `v6.0` the constrains file download will be skipped and requirements only for `master` branch without constraints file will be taken.


**So to make short conclusion, it is for fall-back options to not rise an uncaught Exception or to directly exit the script when bad inputs of GitHub variables will be provided and other requirements can be taken.**

This was the idea behind if it is bad or should be anything else I am open to this.

Reply:
Okay, sorry. The IDF `release/v5.2` is already on GitHub but the idea will be the same.

And thank you for review, I have discovered a mistake in `_download_branch_requirements` function thanks to this.